optaplanner-examples/src/main/java/org/optaplanner/examples/nurserostering/optional/score/NurseRosteringConstraintProvider.java

```diff
@@ -96,9 +99,13 @@ Constraint minimumAndMaximumNumberOfAssignments(ConstraintFactory constraintFact
         return constraintFactory.from(MinMaxContractLine.class)
                 .filter(minMaxContractLine -> minMaxContractLine.getContractLineType() == ContractLineType.TOTAL_ASSIGNMENTS &&
                         minMaxContractLine.isEnabled())
-                .join(ShiftAssignment.class, Joiners.equal(ContractLine::getContract, ShiftAssignment::getContract))
-                .groupBy((line, shift) -> shift.getEmployee(), (line, shift) -> line, ConstraintCollectors.countBi())
-                .filter((employee, contract, shiftCount) -> employee != null && contract.isViolated(shiftCount))
+                .join(constraintFactory.from(ShiftAssignment.class)
+                        .filter(shift -> shift.getEmployee() != null),
+                        Joiners.equal(ContractLine::getContract, ShiftAssignment::getContract))
+                .groupBy((line, shift) -> shift.getEmployee(),
+                        (line, shift) -> line,
+                        ConstraintCollectors.countBi())
+                .filter((employee, contract, shiftCount) -> contract.isViolated(shiftCount))
```

Comment:
This should be faster. I will benchmark all of these changes once your original PR no longer shows score corruptions.